### PR TITLE
Use more points when drawing a circle

### PR DESCRIPTION
**Pull request by Xoronic**
## Problem
I ran the examples in the development branch, and I noticed that the circles looked off. 

## Proposed solution
I wrote a quick Python3 script to generate the array with circlepoints and updated the points.

## Screenshots
This was the pulsing circle example before:
circle_before

This is the pulsing circle example now:
circle_after

## Python Script to generate the array.
```python
import math
from math import pi


def points_on_circumference(center=(0, 0), r=50, n=100):
    n -= 1
    return [
        (
            center[0]+(math.cos(2 * pi / n * x) * r),  # x
            center[1] + (math.sin(2 * pi / n * x) * r)  # y

        ) for x in range(0, n + 1)]


number_of_points = 64
points = points_on_circumference(center=(0,0),r=1, n=number_of_points)

print("const CIRCLE_POINTS: [Vector; %i] = [" % number_of_points)

for point in points:
    print("    Vector { x: %s, y: %s }," % (point[0], point[1]))

print("];")
```

## API changes
None.

### CODE REVIEW

**Comment by ryanisaacg** [comment]
@Xoronic It was not supposed to be weird, thanks for the PR! I think it has to do with the ordering of the vertices causing the very wrong circle in the before picture. I appreciate you embedding the Python script as well!

**Comment by Xoronic** [comment]
I was not sure how many points to generate. 64 points seemed a good balance between efficiency and appearance... What do you think?

**Comment by ryanisaacg** [comment]
It works for now, I hope to come up with a way to allow users to adjust it in the future.

**Comment by Byter09** [comment]
Hey thanks for updating that. I was wondering what was wrong with it.

Now that thats fixed, how about something that generates enough points so the circle always looks round? I'm not quite sure how CPU heavy this is, but I planned this as a new object (you know... the new ones like Line and Triangle), with its vertices always having a max. distance of 2 pixel. Or have it with an optional parameter, setting the number when its given and automating it when not.

Is this too much?

**Comment by Xoronic** [comment]
Maybe it would be more efficient to generate a very long array of points. If the circle is very small, the draw function could just skip values from the array.

**Comment by Xoronic** [comment]
We should do some tests to see how cpu heavy it is.

**Comment by Xoronic** [comment]
Also, if the points are calculated live, we should only calculate a quarter of the circle. We can mirror the other 4 quarters :)

**Comment by ryanisaacg** [comment]
That could definitely work! I'd be interested in seeing it.

**Comment by Byter09** [comment]
Oh that's clever. Saves a lot of time.

So in conclusion, we only need a quarter and vertices have to be 1 to 2 pixel apart to appear smooth. What's the vertices to radius ratio with that? Imagine a circle so big that it looks like a line when viewing the edge. What about that?

**Comment by Xoronic** [comment]
> vertices have to be 1 to 2 pixel apart to appear smooth

r: radius in pixels
n: number of total vertices
L: amount of borderpixels for one edge (should be smaller than 2.5)

untitled

So any angle of **`2.5 / r`** radians or smaller should be sufficient.

So that would result in 
untitled


And only a 4th of that should be calculated.
untitled

**Comment by Xoronic** [comment]
I could write the code for that

**Comment by Byter09** [comment]
At least it is linear. And I love that you took the time to include images of all the maths.. Haha.

If we now solved the problem of "how many points" and where exactly they are, it should be trivial to write that algorithm, right? But, if you can, try to benchmark it or simply find some values that make it take longer, so that we may find edge cases. Whatever the results, I love it. This is exactly what I wanted to do next together with quadrilaterals.

**Comment by Xoronic** [comment]
> EDIT: Oh and I'm curious: How big would a circle have to be, so that no matter where you look at its edge, it would look like a straight line? This would definitely be our maximum r, and therefore the maximum amount of points to generate.

Hmm, interesting problem. That would depend on the size of the window.. You mean something like this, right? 


example

**Comment by Byter09** [comment]
Exactly.

Here's an example:
wH: Window height
wW: Window width
x > wW (circle only right because I'm simplyfing for the example)

Let's take a circle at position (x, wH / 2) and radius x - we/2 , resulting in its edge always being in the middle of the screen. How far to the right (and therefore how big its radius) do we have to move, so that the resulting presentation on the screen could be approximated with a `Line`. (Or if you want to fill the area on the right, a `Triangle`)


I hope this makes sense but you got the idea...

**Comment by Xoronic** [comment]
Pfff, just a couple pages of math :S , but I think I solved the simple case you described above :)

I calculated the intersectionpoints of the circle and the line `y=h`

 `x`: the x-values of the two intersectionpoints (in pixels).
`p`: the x-value of the center of the circle (in pixels).
`w`: the width of the window (in pixels).
`h`: the height of the window (in pixels).

solution


untitled

The value we are interested in is the difference (in the x-dimension) between the intersectionpoint and the center of the window. That value should be smaller than 1.5 (less than one pixel).

The (in this case) minimum x-value should be subtracted from the x-value of the intersectionpoint that is located between the left and the right border of the window.

As the minimum is at the center of the screen (we placed it there), the value of that is half the window's width.

In conclusion, this must be true:
solution2

above gives

solution3

I'm not sure if this works as a common solution if the circle is located elsewhere.. Replacing `x` with `y` , `w` with `h` and `p` with `q` everywhere should give us the solutions for the `y` values, right? I would say that a circle looks like a straight line if above statement is true for the `x` values AND for the `y` values.

I had some fun tackling this problem, but I am getting the feeling this is premature optimization :P

**Comment by Xoronic** [comment]
Where is the origin located in quicksilver by the way? Top-left?

**Comment by ryanisaacg** [comment]
@Xoronic Yup, top-left.

**Comment by Byter09** [comment]
It probably is. Maybe such a circle generation should be implemented first, then we can benchmark it inside the engine itself and see where things go from there.

But I'm just too curious: Did you actually get a definitive answer for, lets say 1920*1080 as the resolution? How big is that circle in that case? Maybe its not even worth optimzing if the result is not too big (can't check it myself, still busy and only looking at this on my phone).
Sadly this still does not solve the problem of generating way too many vertices...

**Comment by Xoronic** [comment]
Oh, wow. All that work and I just found a way better solution.

**Comment by Xoronic** [comment]
I didn't exactly, actually.. Be right back :)

**Comment by Xoronic** [comment]
Nevermind, I found something very close to the solution but it keep going off track..

**Comment by Xoronic** [comment]
> But I'm just too curious: Did you actually get a definitive answer for, lets say 1920*1080 as the resolution?

a radius of 97,440 pixels

**Comment by Xoronic** [comment]
Two pixels with a radius of 58,340 pixels

**Comment by Byter09** [comment]
Ouch. That's a lot of pixels.
So for that circle it would need 
n > 0.2 * pi * 97440
n would need to be 61223.

That's a ton of vertices where most of it is not even relevant. Holy crap.

**Comment by Xoronic** [comment]
Wouldn't it be possible to draw the circle in the fragmentshader?

**Comment by Xoronic** [comment]
I have not taken a look at the backend implementation yet..

**Comment by Byter09** [comment]
I think that's not even necessary.
Let's say we would write an additional object called a segment, which then could be used to draw any circle... (and probably even fix the open arc issue). All we would need for that to work is
- a segment object
- an algorithm to get the segment of the circle, that is visible on the screen.

This solves the vertices problem, takes basically no CPU time if the circle is fully visible and gives us a new toy to play with.

**Comment by Xoronic** [comment]
A [triangle fan](https://en.wikipedia.org/wiki/Triangle_fan) would be a cool optimalisation as well. But let's not optimize prematurely :P

**Comment by Byter09** [comment]
So. I'm gonna summarize it:
- circle is fully visible inside the area we're drawing to
  - draw it like always
  - (optionally draw it using a segment with 360°?)
- circle fully covers the area we're drawing to
  - simply use a triangle that can fit the area inside (this only works, if the the circle has no special colours on its vertices or anything - a plain circle so to speak)
- circle partly covers the area we're drawing to
  - find the segment of the circle thats visible and draw that one instead
- a segment, that partly covers the area can be optimized like this too, by finding the segment inside the segment thats visible

And, I'm going to be honest with you: I really love tricking the user when it comes to optimizing the result so this is a really fun topic for me 😝

**Comment by Byter09** [comment]
I usually don't write multiple messages, but this is important:
I ment a circular sector, not a segment. Got my terminology wrong there. Just in case you were already working on something..

But I'll also try my best and see how it goes.

**Comment by Xoronic** [comment]
I was just reading your message :)

@ryanisaacg Maybe it is a good idea to create an issue for this circle optimization.
I think the issue could use a [task list](https://help.github.com/articles/about-task-lists/), with the exact points mentioned above.

**Comment by Xoronic** [comment]
I just noticed that I made a false assumption about the CIRCLE_POINTS array. Actually, it was a coincidence that the approach above worked. The array I generated would not work for parts of circles, because it draws everything from (1.0, 0.0) instead of drawing from the center (0.0, 0.0). I could fix that but the `CIRCLE_POINTS` array will be replaced anyway.

**Comment by Byter09** [comment]
Currently working on a `Segment` object. Will see how it goes. :P

**Comment by ryanisaacg** [comment]
I think opening an issue is a great idea @Xoronic
